### PR TITLE
Wip/duration

### PR DESCRIPTION
**Pull request by MilosKozak**

### CODE REVIEW

**Comment by MilosKozak** [comment]
image
image

File: tests/reports.test.js [hunk 1]
```diff
@@ -288,9 +288,9 @@ describe('reports', function ( ) {
     $('.ui-button:contains("Save")').click();
 
     var result = $('body').html();
-    //var filesys = require('fs');
-    //var logfile = filesys.createWriteStream('out.txt', { flags: 'a'} )
-    //logfile.write($('body').html());
+    var filesys = require('fs');
+    var logfile = filesys.createWriteStream('out.txt', { flags: 'a'} )
```
Comment (jasoncalabrese) [inline tests/reports.test.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1066741053/issues/source?bid=2452379&fileBranchId=2886508#l292)

**Comment by MilosKozak** [comment]
@jasoncalabrese your comments/improvements?

**Comment by jasoncalabrese** [comment]
Not sure about turning off fields for the default 'BG Check' event type, I think lots of people don't bother with the event type.  Would need to change the UI to the common types more visible.

For the visualization,I think we should try to make it a little lighter, maybe drop the outlines.  Also  think we should be careful about about introducing more colors and things before getting temp basal and combo boluses in.

**Comment by MilosKozak** [comment]
That's one of the reason to add this, grandparents mostly didn't change eventType

**Comment by MilosKozak** [comment]
With temp basals and combo boluses we'll need even more fields. Without hiding unneeded fields drawer becomes confusing

**Comment by jasoncalabrese** [comment]
yeah, I see that issue too, maybe at first we only show the new fields for new types?  A different UI would encourage better type selection.

File: lib/client/chart.js [hunk 2]
```diff
@@ -187,6 +206,7 @@ function init (client, d3, $) {
     // get the height of each chart based on its container size ratio
     var focusHeight = chart.focusHeight = chartHeight * .7;
     var contextHeight = chart.contextHeight = chartHeight * .2;
+    var basalsHeight = chart.basalsHeight = focusHeight / 4;
```
Comment (jasoncalabrese) [inline lib/client/chart.js]: Codacy Issue found: ['basalsHeight' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129533692/issues/source?bid=2452379&fileBranchId=2886508#l209)

File: lib/client/chart.js [hunk 3]
```diff
@@ -358,6 +384,12 @@ function init (client, d3, $) {
           .attr('transform', 'translate(0,' + chartHeight + ')')
           .call(chart.xAxis2);
 
+        var basalsTransition = chart.basals.transition();
```
Comment (jasoncalabrese) [inline lib/client/chart.js]: Codacy Issue found: ['basalsTransition' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129533692/issues/source?bid=2452379&fileBranchId=2886508#l387)

File: lib/client/chart.js [hunk 4]
```diff
@@ -43,6 +43,11 @@ function init (client, d3, $) {
   var yScale2 = chart.yScale2 = d3.scale.log()
     .domain([utils.scaleMgdl(36), utils.scaleMgdl(420)]);
 
+  var xScaleBasals = chart.xScaleBasals = d3.time.scale().domain(extent);
+
+  var yScaleBasals = chart.yScaleBasals = d3.scale.linear()
```
Comment (jasoncalabrese) [inline lib/client/chart.js]: Codacy Issue found: ['yScaleBasals' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129533692/issues/source?bid=2452379&fileBranchId=2886508#l48)

File: lib/client/chart.js [hunk 5]
```diff
@@ -43,6 +43,11 @@ function init (client, d3, $) {
   var yScale2 = chart.yScale2 = d3.scale.log()
     .domain([utils.scaleMgdl(36), utils.scaleMgdl(420)]);
 
+  var xScaleBasals = chart.xScaleBasals = d3.time.scale().domain(extent);
```
Comment (jasoncalabrese) [inline lib/client/chart.js]: Codacy Issue found: ['xScaleBasals' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129533692/issues/source?bid=2452379&fileBranchId=2886508#l46)

File: lib/client/renderer.js [hunk 6]
```diff
@@ -422,6 +485,116 @@ function init (client, d3) {
...
+        .attr('class', 'tempbasaltext')
+        .style('font-size', 15)
+        .attr('fill', 'white')
+        .attr('text-anchor', 'middle')
+        .attr('dy', '.35em')
+        .attr('x', chart().xScaleBasals(t.mills + times.mins(t.duration).msecs/2))
+        .attr('y', 10)
+        .text((t.percent ? (t.percent > 0 ? '+' : '') + t.percent + '%' : '') + (t.absolute ? t.absolute + 'U' : ''));
+      });
+  }
```
Comment (jasoncalabrese) [inline lib/client/renderer.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129533694/issues/source?bid=2452379&fileBranchId=2886508#l596)

File: lib/client/renderer.js [hunk 7]
```diff
@@ -422,6 +485,116 @@ function init (client, d3) {
...
+      .attr('stroke-width', 2)
+      .attr('fill', 'none')
+      .attr('d', valueline(linedata))
+      .attr('clip-path', 'url(#clip)');
+
+    g.append('path')
+      .attr('class', 'line notempline')
+      .attr('stroke', '#0099ff')
+      .attr('stroke-width', 2)
+      .attr("stroke-dasharray", ("3, 3"))
```
Comment (jasoncalabrese) [inline lib/client/renderer.js]: Codacy Issue found: [Mixed double and single quotes.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129533694/issues/source?bid=2452379&fileBranchId=2886508#l560)

File: lib/client/chart.js (hunk 2 above)
Comment (jasoncalabrese) [inline lib/client/chart.js]: Codacy Issue found: ['basalsHeight' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129544652/issues/source?bid=2452379&fileBranchId=2886508#l209)

File: lib/client/chart.js (hunk 3 above)
Comment (jasoncalabrese) [inline lib/client/chart.js]: Codacy Issue found: ['basalsTransition' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129544652/issues/source?bid=2452379&fileBranchId=2886508#l387)

File: lib/client/chart.js (hunk 4 above)
Comment (jasoncalabrese) [inline lib/client/chart.js]: Codacy Issue found: ['yScaleBasals' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129544652/issues/source?bid=2452379&fileBranchId=2886508#l48)

File: lib/client/chart.js (hunk 5 above)
Comment (jasoncalabrese) [inline lib/client/chart.js]: Codacy Issue found: ['xScaleBasals' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129544652/issues/source?bid=2452379&fileBranchId=2886508#l46)

File: lib/client/renderer.js (hunk 6 above)
Comment (jasoncalabrese) [inline lib/client/renderer.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129544654/issues/source?bid=2452379&fileBranchId=2886508#l596)

File: lib/client/renderer.js (hunk 7 above)
Comment (jasoncalabrese) [inline lib/client/renderer.js]: Codacy Issue found: [Mixed double and single quotes.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129544654/issues/source?bid=2452379&fileBranchId=2886508#l560)

File: lib/plugins/basalprofile.js [hunk 8]
```diff
@@ -6,8 +6,18 @@ function init() {
...
     , label: 'Basal Profile'
     , pluginType: 'pill-minor'
+    , additionalHtml: '<input type="checkbox" id="basals-switch" style="margin:0px 0px 0px 0px;vertical-align: bottom;">'
   };
 
+  basal.htmlInitCode = function htmlInitCode () {
+    $('#basals-switch').change(function switchChart (event) {
+      window.Nightscout.client.chart.basals.attr('display', $('#basals-switch').is(':checked') ? '' : 'none');
+      if (event) {
+        event.preventDefault();
```
Comment (jasoncalabrese) [inline lib/plugins/basalprofile.js]: maybe a `event.stopPropagation()` here would stop the tooltip from showing on toggle?

File: lib/client/chart.js (hunk 2 above)
Comment (jasoncalabrese) [inline lib/client/chart.js]: Codacy Issue found: ['basalsHeight' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129704314/issues/source?bid=2452379&fileBranchId=2886508#l209)

File: lib/client/chart.js (hunk 3 above)
Comment (jasoncalabrese) [inline lib/client/chart.js]: Codacy Issue found: ['basalsTransition' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129704314/issues/source?bid=2452379&fileBranchId=2886508#l387)

File: lib/client/chart.js (hunk 4 above)
Comment (jasoncalabrese) [inline lib/client/chart.js]: Codacy Issue found: ['yScaleBasals' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129704314/issues/source?bid=2452379&fileBranchId=2886508#l48)

File: lib/client/chart.js (hunk 5 above)
Comment (jasoncalabrese) [inline lib/client/chart.js]: Codacy Issue found: ['xScaleBasals' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129704314/issues/source?bid=2452379&fileBranchId=2886508#l46)

File: lib/client/renderer.js (hunk 6 above)
Comment (jasoncalabrese) [inline lib/client/renderer.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129704316/issues/source?bid=2452379&fileBranchId=2886508#l596)

File: lib/client/renderer.js (hunk 7 above)
Comment (jasoncalabrese) [inline lib/client/renderer.js]: Codacy Issue found: [Mixed double and single quotes.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129704316/issues/source?bid=2452379&fileBranchId=2886508#l560)

File: lib/client/chart.js (hunk 2 above)
Comment (jasoncalabrese) [inline lib/client/chart.js]: Codacy Issue found: ['basalsHeight' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129921803/issues/source?bid=2452379&fileBranchId=2886508#l209)

File: lib/client/chart.js (hunk 3 above)
Comment (jasoncalabrese) [inline lib/client/chart.js]: Codacy Issue found: ['basalsTransition' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129921803/issues/source?bid=2452379&fileBranchId=2886508#l387)

File: lib/client/chart.js (hunk 4 above)
Comment (jasoncalabrese) [inline lib/client/chart.js]: Codacy Issue found: ['yScaleBasals' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129921803/issues/source?bid=2452379&fileBranchId=2886508#l48)

File: lib/client/chart.js (hunk 5 above)
Comment (jasoncalabrese) [inline lib/client/chart.js]: Codacy Issue found: ['xScaleBasals' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129921803/issues/source?bid=2452379&fileBranchId=2886508#l46)

File: lib/client/renderer.js [hunk 9]
```diff
@@ -422,6 +486,116 @@ function init (client, d3) {
...
+      g.append('text')
+        .attr('class', 'tempbasaltext')
+        .style('font-size', 15)
+        .attr('fill', 'white')
+        .attr('text-anchor', 'middle')
+        .attr('dy', '.35em')
+        .attr('x', chart().xScaleBasals(t.mills + times.mins(t.duration).msecs/2))
+        .attr('y', 10)
+        .text((t.percent ? (t.percent > 0 ? '+' : '') + t.percent + '%' : '') + (t.absolute ? t.absolute + 'U' : ''));
+      });
```
Comment (jasoncalabrese) [inline lib/client/renderer.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129921805/issues/source?bid=2452379&fileBranchId=2886508#l597)

File: lib/client/renderer.js [hunk 10]
```diff
@@ -422,6 +486,116 @@ function init (client, d3) {
...
+      .attr('stroke', '#0099ff')
+      .attr('stroke-width', 2)
+      .attr('fill', 'none')
+      .attr('d', valueline(linedata))
+      .attr('clip-path', 'url(#clip)');
+
+    g.append('path')
+      .attr('class', 'line notempline')
+      .attr('stroke', '#0099ff')
+      .attr('stroke-width', 2)
```
Comment (jasoncalabrese) [inline lib/client/renderer.js]: Codacy Issue found: [Mixed double and single quotes.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129921805/issues/source?bid=2452379&fileBranchId=2886508#l561)

File: lib/client/chart.js (hunk 2 above)
Comment (jasoncalabrese) [inline lib/client/chart.js]: Codacy Issue found: ['basalsHeight' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129941902/issues/source?bid=2452379&fileBranchId=2886508#l209)

File: lib/client/chart.js (hunk 3 above)
Comment (jasoncalabrese) [inline lib/client/chart.js]: Codacy Issue found: ['basalsTransition' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129941902/issues/source?bid=2452379&fileBranchId=2886508#l387)

File: lib/client/chart.js (hunk 4 above)
Comment (jasoncalabrese) [inline lib/client/chart.js]: Codacy Issue found: ['yScaleBasals' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129941902/issues/source?bid=2452379&fileBranchId=2886508#l48)

File: lib/client/chart.js (hunk 5 above)
Comment (jasoncalabrese) [inline lib/client/chart.js]: Codacy Issue found: ['xScaleBasals' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129941902/issues/source?bid=2452379&fileBranchId=2886508#l46)

File: lib/client/renderer.js (hunk 9 above)
Comment (jasoncalabrese) [inline lib/client/renderer.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129941904/issues/source?bid=2452379&fileBranchId=2886508#l597)

File: lib/client/renderer.js (hunk 10 above)
Comment (jasoncalabrese) [inline lib/client/renderer.js]: Codacy Issue found: [Mixed double and single quotes.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1129941904/issues/source?bid=2452379&fileBranchId=2886508#l561)

**Comment by jasoncalabrese** [comment]
huge step forward thanks @MilosKozak

**Pull request merged**